### PR TITLE
Introduced Formatter in libzim

### PR DESCRIPTION
**Pull request by ShaopengLin**
Hi! I am a GSOC student this year and its my first PR here.

Changes:

- Added Formatter Class in src/tools.h. Revised comments to fit our purpose and changed the code from openzim/zim-tools#158 to be more standard.
- Replaced majority of usages of (o)stringstream with Formatter with a few exceptions
- Removed <sstream> imports that are no longer necessary to improve compilation time.

Discussion needed:
- I am not sure if we want the class to be public facing? If no, then the createZimExample.cpp should not be changed. The user would be confused what Formatter is.
- Classes with their own operator<< defined with std::ostream (e.g. blob.h) and functions that uses more complex features of stringstreams (e.g. class CapturedStderr) in my opinion should not need changes. 
    - Making ourselves std::ostream goes against our goal to keep things in a stringstream.  
    - Having the same featrues of stringstreams will likely mean inheritance.

Fix Issue #432

### CODE REVIEW

**Review by mgautierfr — CHANGES_REQUESTED** [review]
Thanks for the PR.

Few things to change.
I'm in favor of using it in `archive.cpp` and `debug.h` too.

**Comment by ShaopengLin** [comment]
I am a little against using it in test/archive. The `CapturedStderr` Class is using rdbuf to read the stream buffer which our class is intended to hide. It makes the class too coupled. If we want to replace all uses of (o)stringstream with Formatter, we should be using inheritance instead. When they need specific features of stringstreams I think it is best they just use the original classes.

File: src/writer/creator.cpp [hunk 1]
```diff
@@ -213,10 +212,10 @@ namespace zim
       auto existing_dirent_it = data->dirents.find(&tmpDirent);
 
       if (existing_dirent_it == data->dirents.end()) {
-        std::ostringstream ss;
-        ss << "Impossible to alias C/" << targetPath << " as C/" << path << std::endl;
-        ss << "C/" << targetPath << " doesn't exist." << std::endl;
-        throw InvalidEntry(ss.str());
+        Formatter ss;
+        ss << "Impossible to alias C/" << targetPath << " as C/" << path << '\n';
```
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/writer/creator.cpp:216]: Will define an overloaded operator to handle std::endl. Misunderstood the `os.widen` on the [std::endl doc](https://en.cppreference.com/w/cpp/io/manip/endl)and thought it was platform independent.

File: src/file_reader.cpp [hunk 2]
```diff
@@ -84,7 +84,7 @@ char MultiPartFileReader::read(offset_t offset) const {
     s << " - physical local offset is " << physical_local_offset.v << "\n";
```
Comment (ShaopengLin) [inline src/file_reader.cpp:84]: Should I change '\n's used with Formatter to std::endl as well? Or do these server a specific purpose?

**Comment by ShaopengLin** [comment]
Also, I think we should place the class in include/zim/tools.h. After installation, error.h would not be able to use this class.

File: src/writer/creator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/writer/creator.cpp]: I found some discussion on this: 
- [Whether inserting endl in string stream makes a difference](https://stackoverflow.com/questions/6865398/inserting-endline-into-a-stringstream)
- [Whether flushing in string stream makes a difference](https://stackoverflow.com/questions/59513853/is-stdstringstreamflush-supposed-to-do-anything)

From what I saw, '\n' and std::endl is equivalent and actually does not make any difference when not writing to text files.

**Comment by ShaopengLin** [comment]
Resolved most of the comments and moved to zim/tools.h to ensure installation is correct. Likely need another patch depending on the discussion of std::endl vs '\n' and archive.cpp

File: src/writer/creator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/writer/creator.cpp]: Indeed, let's keep the `\n` then.

File: src/file_compound.cpp [hunk 3]
```diff
@@ -61,11 +61,8 @@ FileCompound::FileCompound(const std::string& filename):
     } catch (...) { }
 
     if (empty())
-    {
-      std::ostringstream msg;
-      msg << "error " << errnoSave << " opening file \"" << filename;
-      throw std::runtime_error(msg.str());
-    }
+      throw std::runtime_error(Formatter() << "error " << errnoSave
+                                           << " opening file \"" << filename);
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/file_compound.cpp:65]: The `"` must be put after `filename`, not before `"error "`.

**Review by mgautierfr — CHANGES_REQUESTED** [review]
Still a small change to do about closing `"`.

I agree with the global code but please rework the commits. Once this PR will be merged, we will not care about the small mistake made during development and fixed during review. We are more interested with a clean git history.

I would suggest 3 commits (but feel free to assume your own pov on this) :
- Fix closing `"`.
- Introduce `Formatter`.
- Use formater

File: src/file_compound.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/file_compound.cpp]: Resolved in latest cleanup

File: src/writer/creator.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ShaopengLin) [reply src/writer/creator.cpp]: Resolved with the latest overloaded operator for std::endl. The cleanup on '\n's vs std::endl could be in future PRs if necessary.

**Comment by ShaopengLin** [comment]
Fixed the compilation issue with __ostream_type from different platforms. @mgautierfr Can you run the CI again?

**Comment by ShaopengLin** [comment]
I am new to codecov. I believe I should just add test cases in the test folder that hit those code changes right?

**Comment by ShaopengLin** [comment]
So these changes were not part of the covered sections. If we see the sentry report from other PRs, the exact sections are shown as missed. I do not think I have enough knowledge on how to reproduce all these error edge cases. @mgautierfr @kelson42

**Comment by mgautierfr** [comment]
> Should we ignore the codecov atm?

Yes, the formater is used a lot to generate error message. Difficult to fully test.

**Review by mgautierfr — APPROVED** [review]
